add type hints to `base.py`

Dr-Blank opened the pull request:
## Description

Support mypy and/or pyright for the `base.py`

related #1296 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

Dr-Blank committed:
add type hints

- base.py passes for pyright and mypy

Dr-Blank committed:
Remove unnecessary inheritance from Playlist
class.

Dr-Blank committed:
Remove unnecessary inheritance from Photo class.

Dr-Blank committed:
Fix issue with casting None values in utils.py

Dr-Blank committed:
Cleanup

Dr-Blank committed:
More Code refctoring with typehints

Dr-Blank committed:
Add datetime module and update function signatures

glensc commented on plexapi/audio.py:
```diff
@@ -2,6 +2,7 @@
 import os
 from pathlib import Path
 from urllib.parse import quote_plus
+from xml.etree.ElementTree import Element
```
perhaps add such imports that only used for typing under TYPE_CHECKING check?

example:
- https://github.com/Taxel/PlexTraktSync/blob/2f1235e127d69491f3d748d54eccdec9504aef62/plextraktsync/queue/Queue.py#L5-L8

this gives back small performance loss when importing more modules

JonnyWong16 replied:
In this case, this should have no impact on performance because `Element` is already imported in a different module.

glensc commented on plexapi/base.py:
```diff
@@ -1,33 +1,48 @@
 # -*- coding: utf-8 -*-
+from __future__ import annotations
```
for consistency, imho this must be added to any file that uses typing

Dr-Blank replied:
I am planning to typehint other files as well and will implement this for their respective PRs if thats fine.

glensc commented on plexapi/base.py:
```diff
@@ -1,33 +1,48 @@
+
 import re
+from typing import Any, Callable, Dict, List, Optional, Set, Type, TypeVar, Union, cast, overload
+import typing
```
double imports:
- from typing ...
- import typing

glensc commented on plexapi/base.py:
```diff
@@ -321,17 +464,24 @@ def findItems(self, data, cls=None, initpath=None, rtag=None, **kwargs):
+            data = next(utils.iterXMLBFS(data, rtag), Element('empty'))
         # loop through all data elements to find matches
-        items = []
+        items: List[Union[PlexObject, PlexObjectT]] = []
```
```suggestion
        items: list[Union[PlexObject, PlexObjectT]] = []
```

afaik can use native list

JonnyWong16 replied:
Not in Python 3.8.

glensc replied:
eeh. why? any reference?

as it works in my project:
- https://github.com/Taxel/PlexTraktSync/blob/598685fe5fc4f3ae7acd7927218f5f9bc7a71352/setup.cfg#L20
- https://github.com/Taxel/PlexTraktSync/blob/598685fe5fc4f3ae7acd7927218f5f9bc7a71352/plextraktsync/util/execx.py#L6

pyupgrade config:
- https://github.com/Taxel/PlexTraktSync/blob/598685fe5fc4f3ae7acd7927218f5f9bc7a71352/.pre-commit-config.yaml#L44-L49

jjlawren replied:
You need `from __future__ import annotations`.

glensc replied:
I think what happens here is that if you import `__annotations__`, you can use `list[]`:
- https://stackoverflow.com/questions/39458193/using-list-tuple-etc-from-typing-vs-directly-referring-type-as-list-tuple-etc/62033243#62033243

unless it's also a generic:
- https://stackoverflow.com/questions/39458193/using-list-tuple-etc-from-typing-vs-directly-referring-type-as-list-tuple-etc/39458225#39458225

Dr-Blank committed:
Implement requested changes

Dr-Blank commented:
Is there any task remaining for this PR?

ReenigneArcher commented on plexapi/base.py:
```diff
@@ -110,12 +198,12 @@ def _buildItemOrNone(self, elem, cls=None, initpath=None):
             return None
 
-    def _buildDetailsKey(self, **kwargs):
+    def _buildDetailsKey(self, **kwargs: Any):
```
```suggestion
    def _buildDetailsKey(self, **kwargs: Optional[Dict[str, Any]]):
```

Shouldn't kwargs be a dictionary?

I'm unsure the type of the value of the dict, so left that as `Any`

Dr-Blank replied:
is a shorthand, `**_` is automatically parsed as `Dict[str, Unknown]`. providing `Any` is for value as key and dict is understood

ReenigneArcher reviewed (commented):
I have some comments/suggestions.

Some of my comments apply in multiple places, but I only made the comment once. For example the kwargs type hint comment.

Note, I am not a maintainer of this project, but a contributor and user of it.

Dr-Blank committed:
document reason for multiple overloads

Dr-Blank committed:
apply requested changes regarding formatting and code refactoring

Dr-Blank committed:
Sort imports

- annotations import in all files using typehints
- typehint specific imports under type_checking condition
- native library imports > installed libs > module import

Dr-Blank commented on plexapi/playlist.py:
```diff
@@ -2,9 +2,10 @@
 
 from plexapi import media, utils
-from plexapi.base import Playable, PlexPartialObject
+from plexapi.base import Playable
```
maybe i should change my PR title to something like support `base.py` for mypy and pyright. as this change is for that, plexobject is changed as those attributes are used in that class in its methods but are defined in its children. other option would be to move those methods to children.

JonnyWong16 commented on plexapi/base.py:
```diff
@@ -40,23 +56,49 @@ class PlexObject:
+    TAG: Optional[str] = None  # xml element tag
+    TYPE: Optional[str] = None  # xml element type
+    key: Optional[str] = None  # plex relative url
+    _INCLUDES = {
```
Moving `_INCLUDES` here causes them to be added to all `PlexObject` in `_buildDetailsKey`. The includes should only be added to the details key for `PlexPartialObjects`.

Dr-Blank replied:
that is something i did not know. fixed by including an empty member on `PlexObject` to satisfy pyright

JonnyWong16 commented on plexapi/media.py:
```diff
@@ -1,13 +1,20 @@
+if TYPE_CHECKING:
+    from xml.etree.ElementTree import Element
+
+
```
Remove extra blank line.

JonnyWong16 reviewed (changes requested):
I have only done a quick review on GitHub. I haven't actually checked out this PR in my code editor to check.

JonnyWong16 commented:
I also created a new branch for `typehints`. I think it would be better to finish type hinting the entire library before merging into `master.

Dr-Blank commented on plexapi/base.py:
```diff
@@ -1,33 +1,49 @@
 import re
 import weakref
 from functools import cached_property
+from typing import (TYPE_CHECKING, Any, Callable, Dict, List, Optional, Set,
+                    Type, TypeVar, Union, cast, overload)
```
default config or have something in mind for this project?

- all single quotes will be converted to double unless "--skip-string-normalization" is passed
- all docstrings whitespaces will be stripped

Dr-Blank commented on plexapi/base.py:
```diff
@@ -82,26 +124,74 @@ def _clean(self, value):
+        elem: Element,
+        cls: Optional[Type[PlexObjectT]] = None,
+        initpath: Optional[str] = None,
+    ):
```
yes, Optional[Union[PlexObjectT, PlexObject]]

Dr-Blank commented on plexapi/base.py:
```diff
@@ -110,12 +200,12 @@ def _buildItemOrNone(self, elem, cls=None, initpath=None):
             or disable each parameter individually by setting it to False or 0.
         """
-        details_key = self.key
+        details_key = str(self.key)
```
good catch, removed casting

Dr-Blank commented on plexapi/base.py:
```diff
@@ -126,22 +216,38 @@ def _buildDetailsKey(self, **kwargs):
                     See all possible `**kwargs*` in :func:`~plexapi.base.PlexObject.fetchItem`.
         """
-        obj = self
+        obj = self  # ? Why is this needed?
```
isn't object address copied here? or am i missing something?
```python
class Car:
    def __init__(self, name: str, color: str) -> None:
        self.name = name
        self.color = color

    def _dont_mutate_me(self) -> None:
        obj = self  # only memory address of self is copied not the object
        obj.name = "foo"  # will mutate self.name
        assert id(obj) == id(self)  # True

    def __str__(self) -> str:
        return f"I am a {self.color} {self.name}."


def main() -> None:
    car = Car(name="Tesla", color="red")
    print(car)  # I am a red Tesla.
    car._dont_mutate_me()
    print(car)  # I am a red foo.


if __name__ == "__main__":
    main()
```

Dr-Blank commented on plexapi/base.py:
```diff
@@ -40,23 +56,49 @@ class PlexObject:
+        server: PlexServer,
+        data: Element,
+        initpath: Optional[str] = None,
+        parent: Optional[PlexObject] = None,
```
mypy complains
`base.py:167: error: Argument "parent" to "PlexObject" has incompatible type "PlexObject"; expected "PlexObjectT | None"  [arg-type]`

not sure why

was thinking no need for generic type as we are not using that generic type again in any other type definition for the function

Dr-Blank commented on plexapi/base.py:
```diff
@@ -126,22 +216,38 @@ def _buildDetailsKey(self, **kwargs):
+    ) -> PlexObject:
+        ...
+
+    def _manuallyLoadXML(self, xml: str, cls: Optional[Type[PlexObjectT]] = None):
```
updated to `Optional[Union[PlexObjectT, PlexObject]]`

Dr-Blank commented on plexapi/base.py:
```diff
@@ -424,22 +576,22 @@ def _checkAttrs(self, elem, **kwargs):
         if attrstr:
             results = [] if results is None else results
-            for child in [c for c in elem if c.tag.lower() == attr.lower()]:
+            for child in filter(lambda c: c.tag.lower() == attr.lower(), elem):
```
not related to typehints, will remove this from the PR

performance improvment as filter returns an iterator and yields value lazily as opposed to creating the full list in memory.

Dr-Blank committed:
apply suggestions

- fix _INCLUDES attr by including an empty dict
- add explicit return types for methods missing them

Dr-Blank committed:
Merge branch 'typehints' into code/typehints

Dr-Blank commented on plexapi/base.py:
```diff
@@ -1054,7 +1067,7 @@ def _reload(self, _autoReload=False, **kwargs):
     def source(self):
         """ Return the source media object for the session. """
-        return self.fetchItem(self._details_key)
+        return self._details_key and self.fetchItem(self._details_key)
```
fetchItems does not accept None

Dr-Blank reviewed (commented):
marked side effect of suggestion

JonnyWong16 commented on plexapi/base.py:
```diff
@@ -231,10 +337,8 @@ def fetchItems(self, ekey, cls=None, container_start=None, container_size=None,
                     fetchItem(ekey, Media__Part__file__startswith="D:\\Movies")
 
         """
-        if ekey is None:
-            raise BadRequest('ekey was not provided')
```
Why is guard clause removed?

Dr-Blank replied:
```py
def fetchItems(
        self,
        ekey: Union[str, List[int]],
        cls: Optional[Type[PlexObjectT]] = None,
        container_start: Optional[int] = None,
        container_size: Optional[int] = None,
        maxresults: Optional[int] = None,
        **kwargs: Any,
    ):
```

`ekey` is never `None` as defined by the method, hence unnecessary check. could typehint to accept `None` but then calling `fetchItems` without passing `ekey` does not make sense for the method, `None` should be handled by the caller imo

JonnyWong16 commented on plexapi/base.py:
```diff
@@ -253,7 +357,9 @@ def fetchItems(self, ekey, cls=None, container_start=None, container_size=None,
 
             data = self._server.query(ekey, headers=headers)
-            subresults = self.findItems(data, cls, ekey, **kwargs)
+            if not data:
+                return []
```
Is this necessary?

Dr-Blank replied:
`.query` can return `None` and `.findItems` does not accept `None`. so it should be handled in some way. returned empty list as `.fetchItems` returns a list

Dr-Blank replied:
on a second thought maybe returning `result` is a better choice instead of empty list? or should raise error? what is the preferred way to handle no date? other choice is to make query raise an error instead of returning `None`

JonnyWong16 commented on plexapi/base.py:
```diff
@@ -393,20 +535,22 @@ def reload(self, key=None, **kwargs):
             raise Unsupported('Cannot reload an object not built from a URL.')
         self._initpath = key
         data = self._server.query(key)
+        if data is None:
+            raise NotFound(f'Unable to find elem: {key=}')
```
Is this necessary?

Dr-Blank replied:
as long as query could return `None`, handling of `None` should be done, it is done here as next step requires subscripting `data`

` self._loadData(data[0])`

JonnyWong16 commented on plexapi/base.py:
```diff
@@ -40,23 +68,29 @@ class PlexObject:
         self._data = data
         self._initpath = initpath or self.key
         self._parent = weakref.ref(parent) if parent is not None else None
-        self._details_key = None
```
Why is this removed?

Dr-Blank replied:
being assigned after few lines as below and not used before that assignment and this removal, hence the statement has no effect

```py
self._details_key = self._buildDetailsKey()
```

JonnyWong16 commented on plexapi/base.py:
```diff
@@ -940,6 +1089,8 @@ def _reload(self, _autoReload=False, **kwargs):
 
         key = self._initpath
         data = self._server.query(key)
+        if not data:
+            raise NotFound(f'Unable to find elem: {key=}')
```
Is this necessary?

Dr-Blank replied:
```py
   for elem in data:  # None not iterable
```

JonnyWong16 commented on plexapi/collection.py:
```diff
@@ -1,18 +1,22 @@
-    ArtMixin, PosterMixin, ThemeMixin,
-    CollectionEditMixins
-)
+from plexapi.mixins import (AdvancedSettingsMixin, ArtMixin,
+                            CollectionEditMixins, HubsMixin, PosterMixin,
+                            RatingMixin, SmartFilterMixin, ThemeMixin)
```
Don't change in this PR.

JonnyWong16 commented on plexapi/library.py:
```diff
@@ -1,19 +1,25 @@
-    MovieEditMixins, ShowEditMixins, SeasonEditMixins, EpisodeEditMixins,
-    ArtistEditMixins, AlbumEditMixins, TrackEditMixins, PhotoalbumEditMixins, PhotoEditMixins
-)
+from plexapi.mixins import (AlbumEditMixins, ArtistEditMixins,
+                            EpisodeEditMixins, MovieEditMixins,
+                            PhotoalbumEditMixins, PhotoEditMixins,
+                            SeasonEditMixins, ShowEditMixins, TrackEditMixins)
```
Don't change in this PR.

JonnyWong16 commented on plexapi/photo.py:
```diff
@@ -1,16 +1,20 @@
-    ArtUrlMixin, ArtMixin, PosterUrlMixin, PosterMixin,
-    PhotoalbumEditMixins, PhotoEditMixins
-)
+from plexapi.mixins import (ArtMixin, ArtUrlMixin, PhotoalbumEditMixins,
+                            PhotoEditMixins, PosterMixin, PosterUrlMixin,
+                            RatingMixin)
```
Don't change in this PR.

JonnyWong16 commented on plexapi/server.py:
```diff
@@ -1,33 +1,40 @@
+from requests.status_codes import _codes as codes
 
-from plexapi import BASE_HEADERS, CONFIG, TIMEOUT, log, logfilter
+# Need these imports to populate utils.PLEXOBJECTS
```
Don't change in this PR. Move imports back to bottom.

Dr-Blank committed:
revert import statements as suggested

Dr-Blank commented:
@JonnyWong16 

can attr `key` be None in any case? if so, for which cases and is `key=None` useful? as I believe all media items must have a key, hence key cannot be `None` for all of those cases. correct me if I am wrong, but if that is the case, a base class for those items should make it `key` compulsory and for objects that have `key` as None should not inherit from this class with attr `key` in it.

I am saying this as it will save a ton of `if self.key is None` checks in methods which actually rely on `key`

Dr-Blank commented:
should I break this PR further down? Maybe on a per function basis? That should make it easier for review as currently this change is maybe so huge that it is daunting to start reviewing, and multiple threads going on for each change.

Dr-Blank commented:
> I also created a new branch for `typehints`. I think it would be better to finish type hinting the entire library before merging into `master`.

I'm considering whether it would be more beneficial to merge it into `master` directly. Here are my thoughts:

1. Even adding type hints to only `base.py` could greatly enhance readability, providing immediate benefits to users and contributors with making new PRs.
1. By the time the entire project is type hinted, both branches may have diverged significantly, potentially leading to merge conflicts. Merging directly could mitigate this issue.
2. I'm planning to break down PRs on a per-function basis. Since type hints are primarily for developers and don't affect code execution, they're unlikely to introduce breaking changes.

@JonnyWong16, I'd appreciate your thoughts on this. Do you think the benefits of direct merging outweigh any potential risks?

either way I will close this PR and start anew since this is too huge an undertaking for review.

C0rn3j commented:
Why is this marked as a draft?

Even if it is incomplete, it should be merged, as it is useful on its own.

Thanks for your work, I was about to do this myself.

> I wonder if you ruled out providing a stub for typeshed?

While I am not OP - All the currently supported versions of Python support typing, I don't see the point of providing a stub.

Dr-Blank commented:
> Why is this marked as a draft?

I was new and made this a very big PR and hence I assume it was a lot of work to review. Hence I marked it draft as it was not moving ahead and I stopped using this library.

miaLDN commented:
>> I wonder if you ruled out providing a stub for typeshed?
> While I am not OP - All the currently supported versions of Python support typing, I don't see the point of providing a stub.

@C0rn3j rn

You mis-read (whoosh).  Any typing is better than none; inline typing the best; stub better than nothing.  Repeating what I already said and you already didn't understand.
